etcdserver/apply.go

```diff
@@ -745,3 +745,10 @@ func compareInt64(a, b int64) int {
 func isGteRange(rangeEnd []byte) bool {
 	return len(rangeEnd) == 1 && rangeEnd[0] == 0
 }
+
+func noSideEffect(r *pb.InternalRaftRequest) bool {
+	if r.Txn != nil {
+		return isTxnReadonly(r.Txn)
```

Comment:
Ah, do you mean `isTxnReadonly()` should check the condition like below?

`GetRequestPut() == nil && GetRequestDeleteRange() == nil`?